### PR TITLE
[WIP] Escape special characters

### LINKED ISSUE #2043: .gitignore not ignoring files with square brackets
Conditions to reproduce: I am trying to `dvc add` a file that contains brackets in the file name, like `[4] hello.png`. The .gitignore file contains the same text as the file name, but it doesn't work because git will interpret the square brackets differently: https://superuser.com/questions/1125249/filenames-having-square-brackets-not-ignored-by-gitignore

A solution is to add a backslash to [ ] , as \[ and \]. I believe the same issue may be happening with '?'

(at least happens on Linux, but I think it extends to other platforms too)

### PR DESCRIPTION
**Pull request by J0**
* [ ] Have you followed the guidelines in our
      [Contributing document](https://dvc.org/doc/user-guide/contributing)?

* [ ] Does your PR affect documented changes or does it add new functionality
      that should be documented? If yes, have you created a PR for
      [dvc.org](https://github.com/iterative/dvc.org) documenting it or at
      least opened an issue for it? If so, please add a link to it.

-----

Fix #2043

### CODE REVIEW

File: dvc/scm/git/__init__.py [hunk 1]
```diff
@@ -133,6 +134,7 @@ def ignore(self, path, in_curr_dir=False):
     @staticmethod
     def _add_entry_to_gitignore(entry, gitignore, ignore_list):
         content = entry
+        re.escape(content)
```
Comment (shcheklein) [inline dvc/scm/git/__init__.py]: @J0 are we sure that re.espace does the right thing? gitignore is not using regexps, it's a special format - https://git-scm.com/docs/gitignore#_pattern_format . Most likely we need to use some special tool or write our own escape function.

**Comment by J0** [comment]
Alright, will write a few tests then write the corresponding regexp during lunch

**Comment by efiop** [comment]
@J0 Btw, maybe projects like `dulwich` or `pygit2` have needed functions already that we could simply reuse?

File: dvc/scm/git/__init__.py [hunk 2]
```diff
@@ -132,7 +133,25 @@ def ignore(self, path, in_curr_dir=False):
...
+                gitignore_entry = re.sub(
+                    "^([!*])", exclamation_or_asterisk, gitignore_entry
+                )
+                escaped = gitignore_entry.translate(
+                    str.maketrans({"[": r"\[", "]": r"\]", "?": r"\?"})
+                )
+
+            return gitignore_entry
+
+        content = escape_gitignore_special_characters(entry)
```
Comment (Suor) [inline dvc/scm/git/__init__.py]: Looks like this function is a noop. `exclamation_or_asterisk()` is not even called. Looks like indentation issue: code starting with `gitignore_entry = re.sub(` should be unindented.

We are also only quoting `*` at the start of the string, which is probably not intended. I suggest using single regex substitute:
```python
content = re.sub(r'(^!|[*?\[\]])', r'\\\1', entry)
```

**Comment by shcheklein** [comment]
@J0 any update on this? should we just close this as inactive?

File: dvc/scm/git/__init__.py (hunk 2 above)
Comment (J0) [reply dvc/scm/git/__init__.py]: Alright. Thanks for the suggestion --  Will look into this tomorrow morning

**Pull request closed without merging**